p2p/src/authenticated/actors/tracker/actor.rs

```diff
@@ -221,54 +256,48 @@ impl<E: Spawner + Rng + Clock + GClock + Metrics, C: Scheme> Actor<E, C> {
 
             // Iterate over peer set and decrement counts
             for peer in set.order.keys() {
-                if let Some(address) = self.peers.get_mut(peer) {
-                    if address.decrement() {
+                if let Some(record) = self.peers.get_mut(peer) {
+                    if record.decrement() {
                         self.peers.remove(peer);
                     }
                 }
             }
         }
     }
 
+    /// Returns a list of peers that we have a known address for and were able to successfully reserve.
     #[allow(clippy::type_complexity)]
     fn handle_dialable(&mut self) -> Vec<(C::PublicKey, SocketAddr, Reservation<E, C>)> {
-        // Collect unreserved peers
-        let available_peers: Vec<_> = self
+        // Collect peers with known addresses
+        let peers: Vec<(C::PublicKey, SocketAddr)> = self
             .peers
-            .keys()
-            .filter(|peer| !self.connections.contains(*peer))
-            .cloned()
+            .iter()
+            .filter_map(|(peer, record)| record.get_address().map(|addr| (peer.clone(), addr)))
             .collect();
 
-        // Iterate over available peers
-        let mut reserved = Vec::new();
-        for peer in available_peers {
-            // Reserve the connection
-            let reservation = match self.reserve(peer.clone()) {
-                Some(reservation) => reservation,
-                None => continue, // can happen if rate limited
-            };
-
-            // Grab address
-            if let Some(address) = self.peers.get(&peer).unwrap().get_address() {
-                reserved.push((peer, address, reservation));
-            };
-        }
-        reserved
+        // Return all peers that we got a reservation for
+        peers
+            .into_iter()
+            .filter_map(|(peer, addr)| self.reserve(peer.clone()).map(|res| (peer, addr, res)))
+            .collect()
     }
 
     fn handle_peer(&mut self, peer: &C::PublicKey, peer_info: PeerInfo<C>) -> bool {
-        // Check if peer is authorized
-        if !self.allowed(peer) {
+        // Ignore our own peer information
```

Comment:
I like the idea of "auth" checks staying as a single helper. Was the intention to still listen for updates for `blocked` folks with this approach?

Reply:
Yes I was thinking we would still listen/gossip their updated information, but seems it might also be fine to re-use the "allowed" function

Reply:
I think we should stop doing "any work for them" as soon as blocked (in fact, I'd be ok to not gossip their IP either to use that bandwidth for non-blocked peers instead).

Reply:
We could probably mark the address as `Unknown` during block and it should all work (as we'll never update to "known" as soon as blocked).

Reply:
Creating a new enum `Blocked` since it should have slightly different behavior. We will gossip `1` in the bitvec for this peer to avoid receiving its `PeerInfo` from peers